File: lib/repositories/nota_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../../models/nota_model.dart';
import '../../models/bon_model.dart';
import '../../core/enums.dart';

class NotaRepository {
  final SupabaseClient _client;

  NotaRepository(this._client);

  Future<List<NotaModel>> getNotas({
    DateTime? startDate,
    DateTime? endDate,
    String? driverQuery,
    String? factoryId,
  }) async {
    var query = _client.from('notas').select('''
          *,
          relation_agents(*, relation_agent_accounts(*)),
          nota_items!left(count)
        ''');

    if (startDate != null) {
      query = query.gte(
        'invoice_date',
        startDate.toIso8601String().split('T')[0],
      );
    }
    if (endDate != null) {
      final nextDay = DateTime.utc(endDate.year, endDate.month, endDate.day + 1);
      query = query.lt('invoice_date', nextDay.toIso8601String().split('T')[0]);
    }

    if (factoryId != null && factoryId.isNotEmpty) {
      final bonIdsWithFactory = await _client
          .from('bons')
          .select('id')
          .eq('factory_id', factoryId);
      final ids = (bonIdsWithFactory as List)
          .map((b) => b['id'] as String)
          .toList();
      if (ids.isNotEmpty) {
        final itemsResponse = await _client
            .from('nota_items')
            .select('invoice_id')
            .inFilter('bon_id', ids);
        final notaIds = (itemsResponse as List)
            .map((i) => i['invoice_id'] as String)
            .toList();
        if (notaIds.isNotEmpty) {
          query = query.inFilter('id', notaIds);
        } else {
          return [];
        }
      } else {
        return [];
      }
    }

    if (driverQuery != null && driverQuery.isNotEmpty) {
      // Need to find invoices that have at least one bon matching the driver
      final bonIdsWithDriver = await _client
          .from('bons')
          .select('id')
          .ilike('driver_name', '%$driverQuery%');

      final ids = (bonIdsWithDriver as List)
          .map((b) => b['id'] as String)
          .toList();

      if (ids.isNotEmpty) {
        // Find nota_ids associated with these bons
        final itemsResponse = await _client
            .from('nota_items')
            .select('invoice_id')
            .inFilter('bon_id', ids);

        final notaIds = (itemsResponse as List)
            .map((i) => i['invoice_id'] as String)
            .toList();
        if (notaIds.isNotEmpty) {
          query = query.inFilter('id', notaIds);
        } else {
          return []; // No invoices found for this driver
        }
      } else {
        return [];
      }
    }

    final response = await query.order('created_at', ascending: false);
    return (response as List).map((json) => NotaModel.fromJson(json)).toList();
  }

  Future<NotaModel> createNota(NotaModel nota, List<String> bonIds) async {
    return _insertNota(nota.toJson(), bonIds);
  }

  Future<NotaModel> _insertNota(
    Map<String, dynamic> data,
    List<String> bonIds,
  ) async {
    // 1. Create Invoice (Nota)
    final response = await _client
        .from('notas')
        .insert(data)
        .select()
        .single();

    final newNota = NotaModel.fromJson(response);

    // 2. Create Items (Nota Items)
    final items = bonIds
        .map((id) => {'invoice_id': newNota.id, 'bon_id': id})
        .toList();

    await _client.from('nota_items').insert(items);

    // 3. Update Bon Status
    await _client
        .from('bons')
        .update({'status': PaymentStatus.tertagih.value})
        .inFilter('id', bonIds);

    return newNota;
  }

  Future<void> updateNota(
    NotaModel nota,
    List<String> currentBonIds,
    List<String> newBonIds,
  ) async {
    // 1. Update Invoice Basic Data
    await _client.from('notas').update(nota.toJson()).eq('id', nota.id);

    // 2. Diff Bon IDs
    final toRemove = currentBonIds
        .where((id) => !newBonIds.contains(id))
        .toList();
    final toAdd = newBonIds.where((id) => !currentBonIds.contains(id)).toList();

    // 3. Handle Additions
    if (toAdd.isNotEmpty) {
      final items = toAdd
          .map((id) => {'invoice_id': nota.id, 'bon_id': id})
          .toList();
      await _client.from('nota_items').insert(items);
      await _client
          .from('bons')
          .update({'status': PaymentStatus.tertagih.value})
          .inFilter('id', toAdd);
    }

    // 4. Handle Removals
    if (toRemove.isNotEmpty) {
      await _client
          .from('nota_items')
          .delete()
          .eq('invoice_id', nota.id)
          .inFilter('bon_id', toRemove);
      await _client
          .from('bons')
          .update({'status': PaymentStatus.belumDibayar.value})
          .inFilter('id', toRemove);
    }
  }

  Future<void> deleteNota(String id) async {
    // 1. Check for payments before deleting
    final payments = await _client
        .from('payments')
        .select('id')
        .eq('invoice_id', id)
        .limit(1);

    if (payments.isNotEmpty) {
      throw Exception('Nota sudah memiliki pembayaran, tidak dapat dihapus.');
    }

    // 2. Get associated bons to revert their status
    final items = await _client
        .from('nota_items')
        .select('bon_id')
        .eq('invoice_id', id);
    final bonIds = (items as List).map((i) => i['bon_id'] as String).toList();

    if (bonIds.isNotEmpty) {
      await _client
          .from('bons')
          .update({'status': PaymentStatus.belumDibayar.value})
          .inFilter('id', bonIds);
    }

    // 3. Delete nota_items then nota
    await _client.from('nota_items').delete().eq('invoice_id', id);
    await _client.from('notas').delete().eq('id', id);
  }

  Future<String?> getNotaIdByBonId(String bonId) async {
    final response = await _client
        .from('nota_items')
        .select('invoice_id')
        .eq('bon_id', bonId)
        .limit(1);
    if ((response as List).isEmpty) return null;
    final id = response[0]['invoice_id'] as String?;
    return (id == null || id.isEmpty) ? null : id;
  }

  Future<NotaModel?> getNotaById(String id) async {
    final response = await _client
        .from('notas')
        .select(
          '*, relation_agents(*, relation_agent_accounts(*)), nota_items!left(count)',
        )
        .eq('id', id)
        .maybeSingle();
    if (response == null) return null;
    return NotaModel.fromJson(response);
  }

  Future<List<BonModel>> getNotaBons(String notaId) async {
    final response = await _client
        .from('nota_items')
        .select(
          'bons(*, bon_deductions(*), factories(name), relation_agents(name))',
        )
        .eq('invoice_id', notaId);

    return (response as List)
        .map((json) => BonModel.fromJson(json['bons']))
        .toList();
  }

  Future<void> updateNotaTotal(String notaId, double totalAmount) async {
    await _client.from('notas').update({'total_amount': totalAmount.toInt()}).eq('id', notaId);
  }

  /// Gabung beberapa bon menjadi satu nota.
  /// [relationAgentId] wajib dikirim jika relasi antar bon berbeda.
  Future<NotaModel> mergeBonsIntoNota(
    List<String> bonIds, {
    String? relationAgentId,
  }) async {
    final ids = bonIds.toSet().toList();
    if (ids.length < 2) {
      throw Exception('Pilih minimal dua bon untuk digabung.');
    }

    final selected = await _client
        .from('bons')
        .select(
          'id,status,total,relation_agent_id,relation_name,driver_name,plate_number',
        )
        .inFilter('id', ids);
    if ((selected as List).length != ids.length) {
      throw Exception('Ada bon yang tidak ditemukan.');
    }
    if (selected.any(
      (b) => b['status'] == PaymentStatus.lunas.value,
    )) {
      throw Exception('Bon LUNAS tidak dapat digabung.');
    }

    // Cek pembayaran pada nota penampung bon terpilih
    final notaItems = await _client
        .from('nota_items')
        .select('invoice_id')
        .inFilter('bon_id', ids);
    final notaIds = ((notaItems as List).map((i) => i['invoice_id'] as String))
        .toSet()
        .toList();
    if (notaIds.isNotEmpty) {
      final payments = await _client
          .from('payments')
          .select('invoice_id')
          .inFilter('invoice_id', notaIds)
          .limit(1);
      if ((payments as List).isNotEmpty) {
        throw Exception(
          'Salah satu nota sudah memiliki pembayaran, tidak dapat digabung.',
        );
      }
    }

    // Hapus nota penuh atau keluarkan bon terpilih dari nota parsial
    final selectedSet = ids.toSet();
    for (final notaId in notaIds) {
      final notaBons = await getNotaBons(notaId);
      final notaBonIds = notaBons.map((bon) => bon.id).toList();
      final allInSelection =
          notaBonIds.isNotEmpty && notaBonIds.every(selectedSet.contains);
      if (allInSelection) {
        await deleteNota(notaId);
      } else {
        final toRemove = notaBonIds.where(selectedSet.contains).toList();
        if (toRemove.isNotEmpty) {
          await _client
              .from('nota_items')
              .delete()
              .eq('invoice_id', notaId)
              .inFilter('bon_id', toRemove);
          await _client
              .from('bons')
              .update({'status': PaymentStatus.belumDibayar.value})
              .inFilter('id', toRemove);
          final remainingTotal = notaBons
              .where((bon) => !selectedSet.contains(bon.id))
              .fold<double>(
                0,
                (sum, bon) => sum + bon.total,
              );
          await updateNotaTotal(notaId, remainingTotal);
        }
      }
    }

    // Tentukan relasi
    final relIds = selected
        .map((b) => b['relation_agent_id'] as String?)
        .where((id) => id != null && id.isNotEmpty)
        .toSet()
        .toList();
    final relNames = selected
        .map((b) => b['relation_name'] as String?)
        .where((name) => name != null && name.isNotEmpty)
        .toSet()
        .toList();

    Map<String, dynamic> body;
    if (relationAgentId != null && relationAgentId.isNotEmpty) {
      body = {'relation_agent_id': relationAgentId};
    } else if (relIds.length == 1) {
      body = {'relation_agent_id': relIds.first};
    } else if (relNames.length == 1) {
      body = {'recipient_name': relNames.first};
    } else {
      throw Exception('Relasi antar bon berbeda. Pilih relasi untuk nota gabungan.');
    }

    final totalAmount = selected.fold<double>(
      0,
      (sum, bon) => sum + ((bon['total'] as num?)?.toDouble() ?? 0),
    );

    final insertData = <String, dynamic>{
      ...body,
      'invoice_number': 'NOTA-${DateTime.now().millisecondsSinceEpoch}',
      'invoice_date': DateTime.now().toIso8601String(),
      'total_amount': totalAmount.toInt(),
      'status': PaymentStatus.tertagih.value,
    };

    return _insertNota(insertData, ids);
  }
}

File: app_express/src/repositories/notaRepository.js
const { applyDateRange, assertNoError } = require('./base');
const { PAYMENT_STATUS, nowInvoiceNumber } = require('../services/calculations');
const relationAgentRepository = require('./relationAgentRepository');
const { notifyChange } = require('../services/realtimeService');

const CHUNK = 80;

// Ambil invoice_id dari banyak bon secara ber-batch agar URL PostgREST
// tidak melebihi batas panjang server.
async function collectInvoiceIdsForBons(supabase, bonIds) {
  const notaIds = new Set();
  for (let i = 0; i < bonIds.length; i += CHUNK) {
    const chunk = bonIds.slice(i, i + CHUNK);
    const items = assertNoError(
      await supabase.from('nota_items').select('invoice_id').in('bon_id', chunk)
    );
    items.forEach((item) => notaIds.add(item.invoice_id));
  }
  return [...notaIds];
}

async function resolveRecipient(supabase, body) {
  if (body.relation_agent_id) {
    const relation = await relationAgentRepository.getRelationAgent(supabase, body.relation_agent_id);
    return {
      relation_agent_id: relation.id,
      recipient_name: null,
      recipient_address: relation.address || null
    };
  }
  return {
    relation_agent_id: null,
    recipient_name: String(body.recipient_name || '').trim().toUpperCase(),
    recipient_address: body.recipient_address ? String(body.recipient_address).trim().toUpperCase() : null
  };
}

async function listNotas(supabase, filters = {}) {
  const buildQuery = (ids) => {
    let q = supabase.from('notas').select('*, nota_items(bon_id, bons(plate_number, driver_name, netto_2, total)), relation_agents(name)');
    q = applyDateRange(q, 'invoice_date', filters.start, filters.end);
    if (ids) q = q.in('id', ids);
    return q.order('created_at', { ascending: false });
  };

  if (!filters.q && !filters.factory_id) {
    return assertNoError(await buildQuery(null));
  }

  let bonIds;
  if (filters.q) {
    const bons = assertNoError(
      await supabase.from('bons').select('id').ilike('driver_name', `%${filters.q}%`)
    );
    bonIds = bons.map((bon) => bon.id);
  } else {
    // Batasi pencarian bon ke rentang tanggal yang sama agar daftar id tetap kecil.
    let bonsQuery = supabase.from('bons').select('id').eq('factory_id', filters.factory_id);
    bonsQuery = applyDateRange(bonsQuery, 'bon_date', filters.start, filters.end);
    const bons = assertNoError(await bonsQuery);
    bonIds = bons.map((bon) => bon.id);
  }
  if (bonIds.length === 0) return [];

  const notaIds = await collectInvoiceIdsForBons(supabase, bonIds);
  if (notaIds.length === 0) return [];

  const notas = [];
  for (let i = 0; i < notaIds.length; i += CHUNK) {
    notas.push(...assertNoError(await buildQuery(notaIds.slice(i, i + CHUNK))));
  }
  notas.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
  return notas;
}

async function searchNotasByRecipient(supabase, recipientName, filters = {}) {
  const name = String(recipientName || '').trim();
  if (!name) return [];

  let query = supabase
    .from('notas')
    .select('*, nota_items(count), relation_agents!left(name)')
    .or(`recipient_name.ilike.%${name}%,relation_agents.name.ilike.%${name}%`);
  query = applyDateRange(query, 'invoice_date', filters.start, filters.end);
  if (filters.status) query = query.eq('status', filters.status);

  return assertNoError(await query.order('invoice_date', { ascending: false }));
}

async function getNota(supabase, id) {
  return assertNoError(
    await supabase.from('notas').select('*, nota_items(count), relation_agents(*, relation_agent_accounts(*))').eq('id', id).single()
  );
}

async function getNotaBons(supabase, notaId) {
  const rows = assertNoError(
    await supabase.from('nota_items').select('bons(*, bon_deductions(*))').eq('invoice_id', notaId)
  );
  return rows.map((row) => row.bons).filter(Boolean);
}

async function getAvailableBonsForNota(supabase, notaId = null) {
  const allBons = assertNoError(
    await supabase.from('bons').select('*, bon_deductions(*), factories(name), relation_agents(name)').order('bon_date', { ascending: false })
  );
  if (!notaId) return allBons.filter((bon) => bon.status === PAYMENT_STATUS.BELUM_DIBAYAR);

  const currentBons = await getNotaBons(supabase, notaId);
  const currentIds = new Set(currentBons.map((bon) => bon.id));
  return allBons.filter((bon) => bon.status === PAYMENT_STATUS.BELUM_DIBAYAR || currentIds.has(bon.id));
}

async function createNota(supabase, body, bonIds) {
  if (!bonIds.length) throw new Error('Pilih minimal satu bon.');
  const recipient = await resolveRecipient(supabase, body);
  if (!recipient.relation_agent_id && !recipient.recipient_name) throw new Error('Pilih Relasi / Agen penerima nota.');
  const selected = assertNoError(await supabase.from('bons').select('id,total,status').in('id', bonIds));
  if (selected.some((bon) => bon.status !== PAYMENT_STATUS.BELUM_DIBAYAR)) {
    throw new Error('Semua bon yang dipilih harus berstatus BELUM_DIBAYAR.');
  }
  const totalAmount = selected.reduce((sum, bon) => sum + Number(bon.total || 0), 0);

  const nota = assertNoError(
    await supabase
      .from('notas')
      .insert({
        invoice_number: nowInvoiceNumber(),
        invoice_date: new Date().toISOString(),
        total_amount: totalAmount,
        status: PAYMENT_STATUS.TERTAGIH,
        relation_agent_id: recipient.relation_agent_id,
        recipient_name: recipient.recipient_name,
        recipient_address: recipient.recipient_address
      })
      .select('*, relation_agents(*, relation_agent_accounts(*))')
      .single()
  );

  assertNoError(
    await supabase.from('nota_items').insert(
      bonIds.map((bonId) => ({ invoice_id: nota.id, bon_id: bonId }))
    )
  );
  assertNoError(await supabase.from('bons').update({ status: PAYMENT_STATUS.TERTAGIH }).in('id', bonIds));
  notifyChange('notas', 'INSERT', nota);
  return nota;
}

async function updateNota(supabase, id, body, newBonIds) {
  if (!newBonIds.length) throw new Error('Pilih minimal satu bon.');
  const recipient = await resolveRecipient(supabase, body);
  if (!recipient.relation_agent_id && !recipient.recipient_name) throw new Error('Pilih Relasi / Agen penerima nota.');
  const nota = await getNota(supabase, id);
  const currentBons = await getNotaBons(supabase, id);
  const currentIds = currentBons.map((bon) => bon.id);
  const toRemove = currentIds.filter((bonId) => !newBonIds.includes(bonId));
  const toAdd = newBonIds.filter((bonId) => !currentIds.includes(bonId));

  if (toAdd.length > 0) {
    const addCandidates = assertNoError(await supabase.from('bons').select('id,status').in('id', toAdd));
    if (addCandidates.some((bon) => bon.status !== PAYMENT_STATUS.BELUM_DIBAYAR)) {
      throw new Error('Bon tambahan harus berstatus BELUM_DIBAYAR.');
    }
  }

  const allSelected = assertNoError(await supabase.from('bons').select('id,total').in('id', newBonIds));
  const totalAmount = allSelected.reduce((sum, bon) => sum + Number(bon.total || 0), 0);

  assertNoError(
    await supabase
      .from('notas')
      .update({
        total_amount: totalAmount,
        status: nota.status,
        relation_agent_id: recipient.relation_agent_id,
        recipient_name: recipient.recipient_name,
        recipient_address: recipient.recipient_address
      })
      .eq('id', id)
  );

  if (toRemove.length > 0) {
    assertNoError(await supabase.from('nota_items').delete().eq('invoice_id', id).in('bon_id', toRemove));
    assertNoError(await supabase.from('bons').update({ status: PAYMENT_STATUS.BELUM_DIBAYAR }).in('id', toRemove));
  }
  if (toAdd.length > 0) {
    assertNoError(await supabase.from('nota_items').insert(toAdd.map((bonId) => ({ invoice_id: id, bon_id: bonId }))));
    assertNoError(await supabase.from('bons').update({ status: PAYMENT_STATUS.TERTAGIH }).in('id', toAdd));
  }

  return getNota(supabase, id);
}

async function mergeBonsIntoNota(supabase, bonIds, relationAgentId = null) {
  const ids = [...new Set((bonIds || []).map((id) => String(id)).filter(Boolean))];
  if (ids.length < 2) throw new Error('Pilih minimal dua bon untuk digabung.');

  const selected = assertNoError(await supabase.from('bons').select('id,status,total,relation_agent_id,relation_name,driver_name,plate_number').in('id', ids));
  if (selected.length !== ids.length) throw new Error('Ada bon yang tidak ditemukan.');
  if (selected.some((bon) => bon.status === PAYMENT_STATUS.LUNAS)) {
    throw new Error('Bon LUNAS tidak dapat digabung.');
  }

  // Cek pembayaran pada nota yang menampung bon-bon terpilih
  const notaItems = assertNoError(
    await supabase.from('nota_items').select('invoice_id, bon_id').in('bon_id', ids)
  );
  const notaIds = [...new Set(notaItems.map((item) => item.invoice_id))];
  if (notaIds.length > 0) {
    const payments = assertNoError(
      await supabase.from('payments').select('invoice_id').in('invoice_id', notaIds).limit(1)
    );
    if (payments.length > 0) {
      throw new Error('Salah satu nota sudah memiliki pembayaran, tidak dapat digabung.');
    }
  }

  // Hapus nota eksisting (penuh) atau keluarkan bon terpilih dari nota (parsial)
  const selectedSet = new Set(ids);
  for (const notaId of notaIds) {
    const notaBons = await getNotaBons(supabase, notaId);
    const notaBonIds = notaBons.map((bon) => bon.id);
    const allInSelection = notaBonIds.length > 0 && notaBonIds.every((id) => selectedSet.has(id));
    if (allInSelection) {
      await deleteNota(supabase, notaId);
    } else {
      const toRemove = notaBonIds.filter((id) => selectedSet.has(id));
      if (toRemove.length) {
        assertNoError(await supabase.from('nota_items').delete().eq('invoice_id', notaId).in('bon_id', toRemove));
        assertNoError(await supabase.from('bons').update({ status: PAYMENT_STATUS.BELUM_DIBAYAR }).in('id', toRemove));
        const remainingTotal = notaBons
          .filter((bon) => !selectedSet.has(bon.id))
          .reduce((sum, bon) => sum + Number(bon.total || 0), 0);
        assertNoError(await supabase.from('notas').update({ total_amount: remainingTotal }).eq('id', notaId));
      }
    }
  }

  // Tentukan relasi: id yang sama → pakai; relation_name sama → pakai nama; selain itu wajib dikirim relation_agent_id
  const relIds = [...new Set(selected.map((bon) => bon.relation_agent_id).filter(Boolean))];
  const relNames = [...new Set(selected.map((bon) => bon.relation_name).filter(Boolean))];
  let body;
  if (relationAgentId) body = { relation_agent_id: relationAgentId };
  else if (relIds.length === 1) body = { relation_agent_id: relIds[0] };
  else if (relNames.length === 1) body = { recipient_name: relNames[0] };
  else throw new Error('Relasi antar bon berbeda. Pilih relasi untuk nota gabungan.');

  return createNota(supabase, body, ids);
}

async function deleteNota(supabase, id) {
  const nota = await getNota(supabase, id);
  const payments = assertNoError(await supabase.from('payments').select('id').eq('invoice_id', id).limit(1));
  if (payments.length > 0) throw new Error('Nota sudah memiliki pembayaran, tidak dapat dihapus.');
  const bons = await getNotaBons(supabase, id);
  const bonIds = bons.map((bon) => bon.id);
  if (bonIds.length > 0) {
    assertNoError(await supabase.from('bons').update({ status: PAYMENT_STATUS.BELUM_DIBAYAR }).in('id', bonIds));
  }
  assertNoError(await supabase.from('nota_items').delete().eq('invoice_id', id));
  assertNoError(await supabase.from('notas').delete().eq('id', id));
  notifyChange('notas', 'DELETE', null, nota);
}

module.exports = {
  createNota,
  deleteNota,
  getAvailableBonsForNota,
  getNota,
  getNotaBons,
  listNotas,
  mergeBonsIntoNota,
  searchNotasByRecipient,
  updateNota
};
